### PR TITLE
Create metrics to track block ingestion and GraphQL query execution

### PR DESCRIPTION
**Pull request by fordN**
This PR creates two new metrics structs for instrumenting the Block Ingestor and the GraphQL server. 

1. `BlockIngestorMetrics` with a gauge vec to track the number of blocks being synced from Ethereum, labeled by networks name. 

2. `GraphQLServiceMetrics` which includes a histogram vec for tracking query execution time of successful queries, labeled by subgraph deployment id.

### CODE REVIEW

File: chain/ethereum/src/block_ingestor.rs [hunk 1]
```diff
@@ -1,9 +1,35 @@
...
+pub struct BlockIngestorMetrics {
+    pub block_number: Box<GaugeVec>,
+}
+
+impl BlockIngestorMetrics {
+    pub fn new<M: MetricsRegistry>(registry: Arc<M>) -> Self {
+        Self {
+            block_number: registry
+                .new_gauge_vec(
+                    String::from("ethereum_block_number"),
```
Comment (Jannis) [inline chain/ethereum/src/block_ingestor.rs]: We've been calling this the chain head in graph-node. Let's stick to that and call this `ethereum_chain_head_number`.

File: chain/ethereum/src/block_ingestor.rs [hunk 2]
```diff
@@ -1,9 +1,35 @@
...
 use web3::types::*;
 
+pub struct BlockIngestorMetrics {
+    pub block_number: Box<GaugeVec>,
+}
+
+impl BlockIngestorMetrics {
+    pub fn new<M: MetricsRegistry>(registry: Arc<M>) -> Self {
+        Self {
+            block_number: registry
```
Comment (Jannis) [inline chain/ethereum/src/block_ingestor.rs]: And let's rename the variable to `chain_head_number` as well.

File: chain/ethereum/src/block_ingestor.rs [hunk 3]
```diff
@@ -1,9 +1,35 @@
...
+                .new_gauge_vec(
+                    String::from("ethereum_block_number"),
+                    String::from("Records the number of the most recent block synced on the Ethereum network"),
```
Comment (Jannis) [inline chain/ethereum/src/block_ingestor.rs]: Shorten this to `Block number of the most recent block synced from Ethereum`?

File: chain/ethereum/src/block_ingestor.rs [hunk 4]
```diff
@@ -1,9 +1,35 @@
...
+                )
+                .unwrap(),
+        }
+    }
+
+    pub fn observe_blocks_synced(&self, network_name: &str, latest_block_number: i64) {
+        self.block_number
+            .with_label_values(vec![network_name].as_slice())
+            .set(latest_block_number as f64);
+    }
```
Comment (Jannis) [inline chain/ethereum/src/block_ingestor.rs]: Let's also call this `set_chain_head_number`?

File: chain/ethereum/src/block_ingestor.rs [hunk 5]
```diff
@@ -115,6 +147,10 @@ where
                                     LogCode::BlockIngestionStatus
                                 };
                                 if distance > 0 {
+                                    self.ingestor_metrics.observe_blocks_synced(
```
Comment (Jannis) [inline chain/ethereum/src/block_ingestor.rs]: This is not the right place IMHO. At this point we have't written the block to our db yet, so as far as our system is concerned, we are not up to date with the chain yet.

Comment (Jannis) [reply chain/ethereum/src/block_ingestor.rs]: I think the right place would be an `inspect` here: https://github.com/graphprotocol/graph-node/pull/1403/files#diff-dc8c393601039a99522fbd9bdd37b8dbR233-R235

File: chain/ethereum/src/block_ingestor.rs [hunk 6]
```diff
@@ -1,9 +1,35 @@
...
 
 use graph::prelude::*;
 use web3::types::*;
 
+pub struct BlockIngestorMetrics {
+    pub block_number: Box<GaugeVec>,
+}
+
+impl BlockIngestorMetrics {
+    pub fn new<M: MetricsRegistry>(registry: Arc<M>) -> Self {
```
Comment (Jannis) [inline chain/ethereum/src/block_ingestor.rs]: We've been able to get away with `Arc<dyn MetricsRegistry>` in a lot of places. This should work here as well; then we don't have to use generics.

File: chain/ethereum/src/block_ingestor.rs [hunk 7]
```diff
@@ -13,21 +39,26 @@ where
...
+    ingestor_metrics: Arc<BlockIngestorMetrics>,
     polling_interval: Duration,
 }
 
 impl<S> BlockIngestor<S>
 where
     S: ChainStore,
 {
-    pub fn new(
+    pub fn new<M>(
```
Comment (Jannis) [inline chain/ethereum/src/block_ingestor.rs]: Same here regarding `Arc<dyn MetricsRegistry>`.

File: server/http/src/server.rs [hunk 8]
```diff
@@ -65,15 +67,17 @@ impl<Q, S> GraphQLServer<Q, S> {
...
+            metrics_registry,
             graphql_runner,
             store,
             node_id,
         }
     }
 }
 
-impl<Q, S> GraphQLServerTrait for GraphQLServer<Q, S>
+impl<M, Q, S> GraphQLServerTrait for GraphQLServer<M, Q, S>
```
Comment (Jannis) [inline server/http/src/server.rs]: Same here regarding `Arc<dyn MetricsRegistry>`.

File: server/http/src/service.rs [hunk 9]
```diff
@@ -11,6 +13,40 @@ use hyper::{Body, Method, Request, Response, StatusCode};
...
+        write!(f, "GraphQLServiceMetrics {{ }}")
+    }
+}
+
+impl GraphQLServiceMetrics {
+    pub fn new<M: MetricsRegistry>(registry: Arc<M>) -> Self {
+        let query_execution_time = registry
+            .new_histogram_vec(
+                format!("subgraph_query_execution_time"),
+                String::from("Measures the execution time for GraphQL queries"),
```
Comment (Jannis) [inline server/http/src/service.rs]: Drop the `Measures the`. I think that's obvious. :)

File: server/http/src/service.rs [hunk 10]
```diff
@@ -225,27 +266,35 @@ where
...
+                                    "GraphQL query served";
+                                    "subgraph_deployment" => sd_id.deref(),
+                                    "query_time_ms" => elapsed.as_millis(),
+                                    "code" => LogCode::GraphQlQuerySuccess,
+                                )
+                            }
+                            Err(ref e) => error!(
+                                logger,
+                                "GraphQL query failed";
+                                "subgraph_deployment" => sd_id.deref(),
```
Comment (Jannis) [inline server/http/src/service.rs]: Is this not also worth measuring the time for?

File: server/http/src/service.rs [hunk 11]
```diff
@@ -11,6 +13,40 @@ use hyper::{Body, Method, Request, Response, StatusCode};
...
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "GraphQLServiceMetrics {{ }}")
+    }
+}
+
+impl GraphQLServiceMetrics {
+    pub fn new<M: MetricsRegistry>(registry: Arc<M>) -> Self {
+        let query_execution_time = registry
+            .new_histogram_vec(
+                format!("subgraph_query_execution_time"),
```
Comment (Jannis) [inline server/http/src/service.rs]: I wonder if we should shorten this to `query_time`, like in the logs.

**Review by Jannis — CHANGES_REQUESTED** [review]

File: chain/ethereum/src/block_ingestor.rs (hunk 6 above)
Comment (That3Percent) [reply chain/ethereum/src/block_ingestor.rs]: Why do we avoid using generics?

Comment (Jannis) [reply chain/ethereum/src/block_ingestor.rs]: I've been trying to avoid them where I could to simplify type/method signatures. Admittedly, I don't understand the drawbacks of either approach.

Comment (leoyvens) [reply chain/ethereum/src/block_ingestor.rs]: The conclusion is that because this is function parameter the best syntax is `fn new(registry: Arc<impl MetricsRegistry>)`

File: server/http/src/service.rs (hunk 11 above)
Comment (fordN) [reply server/http/src/service.rs]: Want to go with the shorter, `query_time`?  
I vacillated between the two options myself and ended up using `query_execution_time`  because I think it's more clear. In the logs we use `query_time_ms`;  I think aside from being useful to express the unit of measurement,  the `_ms` suffix makes the distinction between execution time and time of execution clear.

File: server/http/src/service.rs (hunk 10 above)
Comment (fordN) [reply server/http/src/service.rs]: Definitely. I had questions about how we want to handle this, so I started with the simplest case.  The error case could return early from a bad request and I don't want those times to skew the aggregate numbers.  

How do you feel about keeping `subgraph_query_execution_time` (or `subgraph_query_time`) for successful queries and add another histogram called `subgraph_failed_query_execution_time`?

File: chain/ethereum/src/block_ingestor.rs (hunk 7 above)
Comment (Jannis) [reply chain/ethereum/src/block_ingestor.rs]: dyn -> impl

File: server/http/src/service.rs (hunk 10 above)
Comment (Jannis) [reply server/http/src/service.rs]: `subgraph_query_time` + `subgraph_failed_query_time`? Sounds good to me 😄

File: server/http/src/service.rs (hunk 11 above)
Comment (Jannis) [reply server/http/src/service.rs]: Ok, let's keep the longer version. We'll want to add more detailed metrics for query execution at some point. We'll see about the names again then.

File: chain/ethereum/src/block_ingestor.rs (hunk 6 above)
Comment (fordN) [reply chain/ethereum/src/block_ingestor.rs]: Hmmm.. `new(registry: Arc<impl MetricsRegistry>)` does not compile for `BlockIngestorMetrics` and `GraphQLServiceMetrics`. 

Screen Shot 2019-12-12 at 10 46 59

Btw, the format I started with, using generics, matches all of the existing metrics struct implementations:
  - `SubgraphInstanceManagerMetrics`
  - `SubgraphInstanceMetrics`
  - `HostMetrics`
  - `ProviderEthRpcMetrics`
  - `SubgraphEthRpcMetrics`
  - `BlockStreamMetrics`

I've updated each of these  according to this conversation and they all successfully compile with `new(registry: Arc<impl MetricsRegistry>)`.

File: chain/ethereum/src/block_ingestor.rs (hunk 5 above)
Comment (fordN) [reply chain/ethereum/src/block_ingestor.rs]: I'm not sure I follow how that would work. Also generally, it is tricky because `ingest_blocks()` will get called again for any missing blocks.  

What about observing the `chain_head_number` from the `ChainHeadUpdateListener`?

Comment (Jannis) [reply chain/ethereum/src/block_ingestor.rs]: I like that idea!

File: chain/ethereum/src/block_ingestor.rs (hunk 6 above)
Comment (Jannis) [reply chain/ethereum/src/block_ingestor.rs]: image

Works fine for me?

Comment (Jannis) [reply chain/ethereum/src/block_ingestor.rs]: That compiles over here.

Comment (fordN) [reply chain/ethereum/src/block_ingestor.rs]: But that's without the signature cleanup suggestions you made about the `BlockIngestor` struct right? ;)

File: server/http/src/service.rs [hunk 12]
```diff
@@ -11,6 +13,58 @@ use hyper::{Body, Method, Request, Response, StatusCode};
...
+
+        let failed_query_execution_time = registry
+            .new_histogram_vec(
+                format!("subgraph_failed_query_execution_time"),
+                String::from("Execution time for failed GraphQL queries"),
+                HashMap::new(),
+                vec![String::from("subgraph_deployment")],
+                vec![0.1, 0.5, 1.0, 10.0, 100.0],
+            )
+            .expect("failed to create `subgraph_query_execution_time` histogram");
```
Comment (Jannis) [inline server/http/src/service.rs]: This should say `subgraph_failed_query_execution_time`.

File: store/postgres/src/chain_head_listener.rs [hunk 13]
```diff
@@ -56,6 +68,10 @@ impl ChainHeadUpdateListener {
                                 notification.payload
                             )
                         });
+                    metrics.set_chain_head_number(
```
Comment (Jannis) [inline store/postgres/src/chain_head_listener.rs]: This deserves a comment IMHO

**Review by Jannis — CHANGES_REQUESTED** [review]

File: store/postgres/src/chain_head_listener.rs (hunk 13 above)
Comment (fordN) [reply store/postgres/src/chain_head_listener.rs]: @jannis, I've added a comment describing the action and intent of using set_chain_head_number() here, please confirm the content is clear and then we should be good to merge this.

Comment (Jannis) [reply store/postgres/src/chain_head_listener.rs]: Yup, that works!

**Review by Jannis — APPROVED** [review]
